### PR TITLE
[ENH] automation to generate sklearn estimator index class source (part 1)

### LINKED ISSUE #118: [ENH] automation to generate estimator index for given package or repository - part 1, `scikit-learn`
We should write automation that writes an estimator index class for a given pypi package or repository.

Input: pypi package name and/or import root name, or repository.
Note: pypi package name and import root name can be different, e.g., `scikit-learn` vs `sklearn`.

Output: a python file (or string with the same content), that is a class like `AiodPkg__Sklearn` if the package contains multiple estimators, or one like `AiodPkg__XGBClassifier` if the package contains a single estimator. Perhaps it should also be an option whether to return the "multiple" or multiple packages in the "single" format.

In first instance, in part 1, we should do this *only*  for `scikit-learn` estimator types, we will expand this later.

This should include tests, we can use the existing packages to test the automation.

### PR DESCRIPTION
**Pull request by samipak458**
## Change

This PR adds automation to generate estimator index class source code for scikit-learn (issue #118, part 1).

The new utility generates valid Python source for AIoD model package classes by composing the existing preindex discovery utilities. It supports both:

- multiple-estimator package format (single class with _obj_dict, _type_of_objs, _objs_by_type)
- single-estimator package format (one class per estimator with _obj and per-class tags)

---

## Motivation

Maintaining large static index classes by hand is error-prone and hard to keep in sync with upstream package changes.
This change introduces a reusable generator so index classes can be created programmatically from discovered estimators and type mappings.

---

### 1) Source generator utility

Added generate_sklearn_index_source(...) in the sklearn preindex utilities.

Current interface:
- pypi_package_name="scikit-learn"
- import_root="sklearn"
- class_name=None
- mode="multiple" (or "single")

Behavior:
- discovers estimator locations via existing locator utility
- derives estimator type mapping via existing type generator
- derives reverse index (objects by type) via existing helper
- renders complete Python source string for generated class(es)

### 2) Helper utilities for robust generation

Added internal helpers to:
- sanitize class name suffixes for generated class names
- normalize object-type values to list form where needed

### 3) Public export

Exposed generate_sklearn_index_source from aiod.utils._indexing so callers can import it from the indexing package root.

### 4) Tests

Added focused tests for:
- generated source structure in multiple mode
- executable correctness (generated class dictionaries match utility outputs)
- expected class/object snippets in single mode

---

## Design notes

- The generator intentionally composes existing utilities rather than re-implementing discovery logic.
- Output is generated as a Python source string so callers can:
  - write to a file directly, or
  - use generated content in other automation flows.
- mode="multiple" defaults class naming to AiodPkg__{ImportRoot} unless class_name is explicitly provided.

---

## Validation

Executed:
- pytest -q tests/test_preindex.py

Result:
- 3 passed

---

## Backward compatibility

- No breaking API changes to existing runtime model retrieval.
- New functionality is additive (new utility + export + tests).

---

## Follow-up scope (not included in this PR)

- generalized non-sklearn package support
- richer class naming/config customization
- integration into an end-to-end generate-and-write command flow

---

## Related Issues

Fixes #118

### CODE REVIEW

**Comment by samipak458** [comment]
> Why do you think we would need
> 
> > single-estimator package format (one class per estimator with _obj and per-class tags)
> 
> ?

Hi @EmanAbdelhaleem,
The single-estimator format is not required for the core part. I included it as an **optional generation mode for future flexibility (packages that expose only one relevant estimator, or workflows that prefer one class per artifact)**. But for issue #118, the essential requirement is the multiple-estimator index class generation for scikit-learn. 

I can scope this PR down and remove single-mode if you prefer to keep part 1 minimal.

Looking forward to your response.

**Comment by samipak458** [comment]
Hi @EmanAbdelhaleem,
Following up on my previous message regarding the scope of this PR for issue #118.

I’m happy to simplify it and keep part 1 focused strictly on the multiple-estimator index class generation for scikit-learn, and remove the optional single-estimator mode if that helps move things forward.

Looking forward to your guidance.

CC: @fkiraly
